### ISSUE TITLE
How to reproduce results on CIFAR-10 ResNet-20.

### ISSUE THREAD

**Issue by sofa13** [fdraxler/PyTorch-AutoNEB#3]
Hello,

Really interesting paper and results! I'm looking to reproduce your results on CIFAR-10 ResNet-20. So far I've been able to run the config "cifar10-resnet20.yaml", but limiting minima_cnt: 2 instead of 10, in order to get results more quickly. After running was complete, I used the Evaluate.ipynb to take a look at the resulting graph.p. I found the loss along the path between the two minima was smooth and close to 0, up until the end, where the loss spikes up to 0.4 and back down again (see figure). 

![cifar10-resnet20_2minima](https://user-images.githubusercontent.com/9827185/47518996-7bc58a80-d85a-11e8-9438-ff777a8a3616.png)

However I was expecting a smooth path without spikes connecting the 2 minima as shown Figure 4 in your paper, after 14 cycles. Should I be able to reproduce similar results as Figure 4 using config "cifar10-resnet20.yaml" with 2 minima?

Thanks!

**Comment by sofa13** [comment]
Thanks for your response, makes sense now that I've re-read section 3.2 as you suggested! I was suspecting it might be a matter of calculating more minima, so I also tried with the CNN-12 architecture with 4 minima, which took about 15 hours in total, about 2.5 hours per edge. However all the loss paths had a similar problem with a spike towards the end. 

I've followed your clear steps to remove the connection data since I am short on computation. I'll be running the connection producer hopefully sometime this weekend and can let you know of any updates. Thanks again for your clear explanation, I really appreciate it!

**Comment by fdraxler** [comment]
Now this sounds weird that you systematically find spikes at the end of the path. Can you maybe quickly say which PyTorch version you are using? I only tested with 0.4.0 and there seem to be some changes with buffers (used by batch norm) in ≥ 0.4.1 that can lead to unexpected behaviour with this code, see #2.

**Comment by fdraxler** [comment]
Oh, and some spike is expected, at least for the shallow and narrow architectures. What average loss value do you find?

**Comment by sofa13** [comment]
Yes I came across an issue with the reduce function when using pytorch 0.4.1, and I installed 0.4.0 when I noticed issue #2, so all calculations have been done using pytorch 0.4.0.
Averages over minimum spanning tree:
Minima:  0.5408 ± 0.0111 (train_loss)
Saddles: 1.0500 ± 0.0493 (saddle_train_loss)
I've provided a figure of the loss path below as well.
![cifar10-cnn_4minima](https://user-images.githubusercontent.com/9827185/47959234-db3b3d00-dfb4-11e8-9353-c45eeda42c63.png)

**Comment by sofa13** [comment]
Hello, just a quick update, I was able to run the connection procedure again on config cifar10-resnet20, after removing the connection data. In order to remove all connections, I had to modify the code a bit, otherwise only the last edge would be removed, rather than all edges:
`edges = list(graph.edges())`
`graph.remove_edges_from(edges)`
The new connection found has an average loss of 0.4138 and is still exhibiting the spike towards the end. Not sure if it is an environment problem on my end... I'll keep trying with more minimas.
![cifar10-resnet20-copy](https://user-images.githubusercontent.com/9827185/48305872-60b47500-e500-11e8-80f8-ba0f07226254.png)
Thanks!

**Comment by fdraxler** [comment]
Okay, I haven't seen this behaviour before. Maybe I made changes to the code at some point after running cifar10-resnet20. I am re-running this now in my setup and will see if I have the same problem.

**Comment by sofa13** [comment]
Okay, thanks for looking into that. Yeah that would definitely interest me! Could I ask for the complete connected graph of the ResNet-20, not just the minimum spanning tree? Thanks!

**Comment by sofa13** [comment]
Hey just wanted to follow up about getting those graphs of ResNet-20 :) Thanks!

**Comment by fdraxler** [comment]
Dear Sophie,Sorry for the delay! I am currently traveling and could not check yet if I still have the full graph (they take a lot of space). If I have a good connection, I can upload at least the mst on Friday.
null

**Comment by fdraxler** [comment]
So, I have the following news:
1) DenseNet results are reproducible with the current code (I successfully ran it in the past).
2) ResNet-20 CIFAR-10 (and I guess other ResNets, too) is not reproducible right now, I cannot dig into the bug until Dec 15th (travelling), but I will.
3) In the meantime, ResNet-20 CIFAR-10 MST is available at: https://drive.google.com/drive/folders/1VQvecH3lWntBD5H0VHyYU577DunvkQlb
If you wish for any more MSTs for any paper experiment, please notify me (not all fit on a standard Google Drive, so I will do this upon request).
4) I probably do not have the full graphs due to storage limitations (those paths are huge). If I stumble on a backup of some system, I will notify you.

**Comment by fdraxler** [comment]
There was a typo in the tangent computation for saddle points that is now corrected.
